[Security Solution][Endpoint] Response actions console isolate command

paul-tavares opened the pull request:
## Summary

- Adds a menu item to the Alert details flyout, for Endpoint alerts only, that will launch the Responder console
- Adds a menu item to the Endpoint List and Endpoint details flyout for launching the Responder console
- Adds console's `isolate` command (send a request to isolate endpoint)
- Adds console's `status` command (endpoint status)
- Adds additional render functions to the Endpoint App Test renderer (`renderHook()`, `renderReactQueryHook()`)



> ------------
> ❗ **NOTE:** ❗
> - The `responseActionsConsoleEnabled` experimental flag must be set in order to see these changes
> - The UI being displayed is not yet adjusted to match the latest UX mocks
> ________


![olm-3481-console-response-actions-isolate-command](https://user-images.githubusercontent.com/56442535/167939287-60820175-57e4-4f4b-bceb-fb1f58c1086b.gif)


### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

paul-tavares committed:
Delete the Top Header consoles component

paul-tavares committed:
Adds action to the take action button + initial setup for action handling

paul-tavares committed:
Improve the display of the Response actions menu item

paul-tavares committed:
new endpoint hooks for showing a console ++ get endpoint host info from api

paul-tavares committed:
Show console when action is clicked from alerts

paul-tavares committed:
Create isolate action via API (still need to monitor it)

paul-tavares committed:
Add `refresh: 'wait_for'` to the isolation api when creating the actions

paul-tavares committed:
Add API calls to continuously check for action complete

paul-tavares committed:
Fix bug in Endpoint list TableRowActions that prevented call to `onClick`

paul-tavares committed:
Add console action to endpoint list

paul-tavares committed:
new general components for Agent/Isolation status and Policy Status

paul-tavares committed:
new hook for getting endpoint pending actions ++ new status console command

paul-tavares committed:
retrieve and show isolation pending data on the status action

paul-tavares committed:
new hook for isolating a host

paul-tavares committed:
Refactor endpoint console commands and action response handlers

paul-tavares committed:
Delete old console command service and add function to get list of commands

paul-tavares committed:
Adjustments to the console menu item label

paul-tavares committed:
Add generics for CommandDefinition `meta` to Console's CommandDefinition, Command and CommandExecutionComponentProps

paul-tavares committed:
Adjustments to the generic types for Console

paul-tavares committed:
Merge remote-tracking branch 'upstream/main' into task/olm-3481-response-ations-console-isolate-command

paul-tavares committed:
Refactor Console command `setStore()` to take in a function that receives current state and returns new state

paul-tavares committed:
Renaming several endpoint hooks

paul-tavares committed:
new hook to get action details + isolate command checks if action is complete

paul-tavares committed:
Improve generic types for Command Execution component

paul-tavares committed:
Fix endpoint details menu item not opening console

paul-tavares committed:
Delete devConsole + fix types

paul-tavares committed:
Fixed type errors in use of generics (but uses `any` :(

paul-tavares committed:
Improve status command to ensure we don't send duplicate fetch's

paul-tavares committed:
revert change to tsconfig.base.json

paul-tavares committed:
Fix i18n duplicate key

paul-tavares committed:
Delete unused component from console

paul-tavares committed:
Fix tests for `TakeActionDropdown` of alerts flyout

paul-tavares committed:
Fix console manager tests

paul-tavares committed:
Fix tests for Host last seen + updated snapshots

paul-tavares committed:
Updated snapshots for side panel tests in timeline

paul-tavares committed:
Refactor endpoint hooks to support overriding the `queryKey`

paul-tavares committed:
Refactor the `status` action to ensure ReactQuery data for the same endpoint id does not impact prior commands

paul-tavares committed:
Fix test in last event time

paul-tavares committed:
Merge remote-tracking branch 'upstream/main' into task/olm-3481-response-ations-console-isolate-command

paul-tavares committed:
Fix i18n files

paul-tavares committed:
remove commented out old code

paul-tavares committed:
Fix types (I think) for CommandExecutionComponentProps

paul-tavares committed:
Merge remote-tracking branch 'upstream/main' into task/olm-3481-response-ations-console-isolate-command

paul-tavares committed:
New host isolation api mocks + improvements to ConsoleManager mocks

paul-tavares committed:
EndpointActionGenerator `generateActionDetails()` method

paul-tavares committed:
Test for isolate action + mocks for response actions

paul-tavares committed:
Action Detail api changes to report `wasSuccessful` and `errors`

paul-tavares committed:
Isolate command - store action details once complete and show errors if any

paul-tavares committed:
Tests for error conditions on isolate command

paul-tavares committed:
additional tests for `getActionCompletionInfo()` to cover new attributes

paul-tavares committed:
Adjust console type for `setState()`

paul-tavares committed:
Merge remote-tracking branch 'upstream/main' into task/olm-3481-response-ations-console-isolate-command

paul-tavares committed:
Fix console "Take action" to ensure it is visible (but disabled) even for non-endpoint alerts

paul-tavares committed:
Initial set of tests for Console menu item in "Take Action" menu

paul-tavares committed:
Restructure tests for endpoint "take action" menu items

paul-tavares committed:
Tests for Launch responder action

elasticmachine commented:
Pinging @elastic/security-onboarding-and-lifecycle-mgt (Team:Onboarding and Lifecycle Mgt)

paul-tavares commented:
@ashokaditya , @dasansol92 - I'm still working on some additional tests for code coverage, but code is ready for review/feedback.


________

@elastic/awp-platform,  @elastic/security-threat-hunting-explore, @elastic/security-threat-hunting-investigations, 

FYI:  You all got pinged because of impact mainly to test files where snapshots had to be updated.

paul-tavares committed:
tests for EndpointAgentAndIsolationStatus component

paul-tavares committed:
Merge remote-tracking branch 'upstream/main' into task/olm-3481-response-ations-console-isolate-command

paul-tavares committed:
Fix tests for Action Details by id

paul-tavares committed:
Fix `use_insert_timeline` tests

I don't know why this fixes the problem, but without it, I get:

```
TypeError: Cannot read properties of undefined (reading 'mockTimelineModel')

      37 | jest.mock('../../../common/hooks/use_selector', () => ({
      38 |   useShallowEqualSelector: jest.fn().mockReturnValue({
    > 39 |     timelineTitle: mockTimelineModel.title,
         |                    ^
      40 |     timelineSavedObjectId: mockTimelineModel.savedObjectId,
      41 |     graphEventId: mockTimelineModel.graphEventId,
      42 |     timelineId: mockTimelineModel.id,
```

opauloh reviewed (approved):
awp-platform changes LGTM

paul-tavares committed:
Fix timeline body actions tests

Not sure why it was failing, but without this fix, I was getting:

```
TypeError: Cannot read properties of undefined (reading 'mockTimelineModel')

      24 | }));
      25 | jest.mock('../../../../../common/hooks/use_selector', () => ({
    > 26 |   useShallowEqualSelector: jest.fn().mockReturnValue(mockTimelineModel),
         |                                                      ^
      27 | }));
      28 | jest.mock(
      29 |   '../../../../../detections/components/alerts_table/timeline_actions/use_investigate_in_timeline',

      at x-pack/plugins/security_solution/public/timelines/components/timeline/body/actions/index.test.tsx:26:54
```

paul-tavares commented on x-pack/plugins/security_solution/public/timelines/components/timeline/body/actions/index.test.tsx:
```diff
@@ -87,6 +86,10 @@ const defaultProps = {
 
 describe('Actions', () => {
+  beforeAll(() => {
+    (useShallowEqualSelector as jest.Mock).mockReturnValue(mockTimelineModel);
```
I don't know why this problem was happening, but I had to change the way the `useShallowEqualSelector` mock and set it here instead of providing `jest.mock()` a factory function. The error I was getting (including in CI) was:

```
TypeError: Cannot read properties of undefined (reading 'mockTimelineModel')

      24 | }));
      25 | jest.mock('../../../../../common/hooks/use_selector', () => ({
    > 26 |   useShallowEqualSelector: jest.fn().mockReturnValue(mockTimelineModel),
         |                                                      ^
      27 | }));
      28 | jest.mock(
      29 |   '../../../../../detections/components/alerts_table/timeline_actions/use_investigate_in_timeline',

      at x-pack/plugins/security_solution/public/timelines/components/timeline/body/actions/index.test.tsx:26:54
```

paul-tavares commented on x-pack/plugins/security_solution/public/cases/components/use_insert_timeline/index.test.tsx:
```diff
@@ -7,7 +7,7 @@
 import { renderHook } from '@testing-library/react-hooks';
 
-import { mockTimelineModel } from '../../../common/mock';
+import { mockTimelineModel } from '../../../common/mock/timeline_results';
```
Without this change, the tests here were failing (CI and locally) with:


```
TypeError: Cannot read properties of undefined (reading 'mockTimelineModel')

      37 | jest.mock('../../../common/hooks/use_selector', () => ({
      38 |   useShallowEqualSelector: jest.fn().mockReturnValue({
    > 39 |     timelineTitle: mockTimelineModel.title,
         |                    ^
      40 |     timelineSavedObjectId: mockTimelineModel.savedObjectId,
      41 |     graphEventId: mockTimelineModel.graphEventId,
      42 |     timelineId: mockTimelineModel.id,
```

paul-tavares committed:
Fix export

paul-tavares committed:
Merge remote-tracking branch 'upstream/main' into task/olm-3481-response-ations-console-isolate-command

paul-tavares committed:
Adjust look of EndpointPolicyStatus

paul-tavares committed:
Rename component for clarity

paul-tavares committed:
Tests for EndpointAppliedPolicyStatus component

paul-tavares committed:
Delete un-used action details service

paul-tavares committed:
New methods in `app_context_render` mock for hooks

paul-tavares committed:
Tests for `useGetActionDetails()` hook

paul-tavares committed:
Tests for `useGetEndpointDetails()` hook

paul-tavares committed:
Generator and mock for Endpoint pending action summary api/response

paul-tavares committed:
Tests for `useGetEndpointPendingActionsSummary()` hook

paul-tavares committed:
Merge remote-tracking branch 'upstream/main' into task/olm-3481-response-ations-console-isolate-command

paul-tavares committed:
Merge remote-tracking branch 'upstream/main' into task/olm-3481-response-ations-console-isolate-command

paul-tavares commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into task/olm-3481-response-ations-console-isolate-command

paul-tavares commented:
buildkite test this

paul-tavares commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into task/olm-3481-response-ations-console-isolate-command

dasansol92 reviewed (commented):
Left few comments but this looks 🔥 A lot of good code here!! Let me know if something is not 
quite clear.

ashokaditya commented on x-pack/plugins/security_solution/public/timelines/components/timeline/body/actions/index.test.tsx:
```diff
@@ -13,16 +13,15 @@ import { Actions } from '.';
-jest.mock('../../../../../common/hooks/use_selector', () => ({
-  useShallowEqualSelector: jest.fn().mockReturnValue(mockTimelineModel),
-}));
+jest.mock('../../../../../common/hooks/use_selector');
```
So I ran this test with 
```
jest.mock('../../../../../common/hooks/use_selector', () => ({
  useShallowEqualSelector: jest.fn(),
}));
```
and it worked fine.

paul-tavares replied:
🤷‍♂️

ashokaditya commented on x-pack/plugins/security_solution/public/management/components/endpoint_applied_policy_status/endpoint_applied_policy_status.tsx:
```diff
@@ -0,0 +1,85 @@
+/**
+ * Displays the status of an applied policy on the Endpoint (using the information provided
+ * by the endpoint in the Metadata document `Endpoint.policy.applied`.
+ * By default, the policy status is displayed as plain text, however, that can be overriden
```
nit: `overridden`

ashokaditya commented on x-pack/plugins/security_solution/public/management/components/endpoint_agent_and_isolation_status/endpoint_agent_and_isolation_status.test.tsx:
```diff
@@ -0,0 +1,47 @@
+    };
+  });
+
+  it('should display only host status when `isIsolated` is undefined', () => {
```
nit: `...display host status only when ...`

ashokaditya commented on x-pack/plugins/security_solution/public/management/components/console/components/command_execution_output.tsx:
```diff
@@ -8,6 +8,7 @@
 import React, { memo, useCallback, useMemo } from 'react';
 import { EuiLoadingChart } from '@elastic/eui';
 import styled from 'styled-components';
+import { CommandExecutionComponentProps } from '../types';
```
Let's state explicitly here and elsewhere that types are imported `import type { CommandExecutionComponentProps }...`

paul-tavares replied:
👍 

_(I wonder why eslint is not catching this - it catches it in some other directories. My guess is that its turned off for some areas fo security solution 🤷‍♂️ )

ashokaditya commented on x-pack/plugins/security_solution/public/common/mock/endpoint/app_context_render.tsx:
```diff
@@ -13,6 +13,14 @@ import { AppDeepLink } from '@kbn/core/public';
+  RenderHookResult,
+} from '@testing-library/react-hooks';
+import { ReactHooksRenderer, WrapperComponent } from '@testing-library/react-hooks/src/types/react';
+import { UseBaseQueryResult } from 'react-query/types/react/types';
```
```suggestion
import types{ UseBaseQueryResult } from 'react-query/types/react/types';
```

ashokaditya commented on x-pack/plugins/security_solution/public/common/mock/endpoint/http_handler_mock_factory.ts:
```diff
@@ -92,6 +92,7 @@ interface RouteMock<R extends ResponseProvidersInterface = ResponseProvidersInte
    */
   id: keyof R;
   method: HttpMethods;
+  /** The API path to match on. This value could can have tokens in the format of `{token_name}` */
```
nit: `...value can have...`

ashokaditya commented on x-pack/plugins/security_solution/public/detections/components/response_actions_console/use_response_actions_console_action_item.tsx:
```diff
@@ -0,0 +1,46 @@
+ */
+
+import React, { useMemo } from 'react';
+import { TimelineEventsDetailsItem } from '@kbn/timelines-plugin/common';
```
```suggestion
import type { TimelineEventsDetailsItem } from '@kbn/timelines-plugin/common';
```

ashokaditya reviewed (commented):
I tested this out locally and the menu items are present and work as expected. I have a couple of spelling nits and some suggestions.

paul-tavares commented on x-pack/plugins/security_solution/public/detections/components/response_actions_console/use_response_actions_console_action_item.tsx:
```diff
@@ -0,0 +1,46 @@
+    if (isResponseActionsConsoleEnabled) {
+      actions.push(
+        <ResponseActionsConsoleContextMenuItem
+          endpointId={isEndpointAlert ? endpointId : ''}
```
I had discussed this before. I think that when someone is looking at the list of alerts and sees options on the menu change between which alert is selected is confusing to the user. Its not clear why the option is no longer there, when it was present in the prior alerts. So this lets it go through and the code in the actual button will disable the option and add a tooltip to it explaining to the user why its disabled. 

The exception to that is Authz - which looks like perhaps I forgot to implement here. Making the change now

paul-tavares commented on x-pack/plugins/security_solution/public/management/components/console/types.ts:
```diff
@@ -54,43 +56,57 @@ export interface CommandDefinition {
-  command: Command;
+export interface CommandExecutionComponentProps<
+  TStore extends object = Record<string, unknown>,
+  TMeta = any
```
I tried and was not able to get it to work in a reliable way. since these are not really used by the Console (they are just "pass-through" to the component that handles the execution of the console command), we really don't care here what its set to. Thus why I turned off the `any` rule at the top of the file

paul-tavares commented on x-pack/plugins/security_solution/public/management/components/console/types.ts:
```diff
@@ -54,43 +56,57 @@ export interface CommandDefinition {
+ */
+export type CommandExecutionComponent<
+  TStore extends object = Record<string, unknown>,
+  TMeta = any
```
same 😞

paul-tavares commented on x-pack/plugins/security_solution/public/management/components/endpoint_console/endoint_response_actions_console_commands.ts:
```diff
@@ -0,0 +1,40 @@
+  return [
+    {
+      name: 'isolate',
+      about: 'Isolate the host',
```
Good point. Yes, they should. Will make the change. Thanks for catching it.

paul-tavares replied:
also - I misspelled the file name 🤦‍♂️

paul-tavares committed:
corrections to spellings + i18n for endpoint commands

paul-tavares committed:
correct file name spelling

paul-tavares commented on x-pack/plugins/security_solution/public/management/components/endpoint_console/status_action.tsx:
```diff
@@ -0,0 +1,179 @@
+
+      // Using a unique `queryKey` here and below so that data is NOT updated
+      // from cache when future requests for this endpoint ID is done again.
+      fetchEndpointDetails({ queryKey: uuidV4() })
```
I guess that wold work as well. I did it this way so that I did not have to `useMemo()` the `queryKey`. This ensures that I call `fetchEndpiontDetails()` only once.

paul-tavares commented on x-pack/plugins/security_solution/public/management/hooks/endpoint/use_get_action_details.ts:
```diff
@@ -0,0 +1,31 @@
+    ...options,
+    queryFn: () => {
+      return http.get<ActionDetailsApiResponse>(
+        resolvePathVariables(ACTION_DETAILS_ROUTE, { action_id: actionId || 'undefined' })
```
good idea. Will do.

paul-tavares commented on x-pack/plugins/security_solution/public/management/hooks/endpoint/use_show_endpoint_response_actions_console.ts:
```diff
@@ -0,0 +1,41 @@
+ */
+
+import { useCallback } from 'react';
+import { getEndpointResponseActionsConsoleCommands } from '../../components/endpoint_console/endoint_response_actions_console_commands';
```
yeah, just did it now 🤦‍♂️

paul-tavares commented on x-pack/plugins/security_solution/server/endpoint/services/actions/utils.ts:
```diff
@@ -96,26 +98,42 @@ export const getActionCompletionInfo = (
+      if (!normalizedAgentResponse.wasSuccessful) {
+        completedInfo.wasSuccessful = false;
+        responseErrors.push(
+          ...(normalizedAgentResponse.errors ? normalizedAgentResponse.errors : [])
```
😁 
I'm spreading an array, so 

```typescript
normalizedAgentResponse.errors ? normalizedAgentResponse.errors : []
```

Will return an array to spread, so I think its ok - right?

paul-tavares committed:
adjustments/corrections from code review

paul-tavares committed:
Merge remote-tracking branch 'upstream/main' into task/olm-3481-response-ations-console-isolate-command

paul-tavares committed:
Merge remote-tracking branch 'origin/task/olm-3481-response-ations-console-isolate-command' into task/olm-3481-response-ations-console-isolate-command

paul-tavares committed:
Apply suggestions from code review

PR review comments from Ash

Co-authored-by: Ashokaditya <1849116+ashokaditya@users.noreply.github.com>

paul-tavares committed:
correct import

dasansol92 reviewed (approved):
🚢 it!

paul-tavares committed:
Change event tests to use mock generator functions

paul-tavares committed:
Fix tests for take action dropdown

paul-tavares committed:
Moved mocks to separate module to avoid calling import of mock from another test

paul-tavares commented on x-pack/plugins/security_solution/public/common/utils/route/index.test.tsx:
```diff
@@ -9,64 +9,36 @@ import { mount } from 'enzyme';
-  pathName: '/',
-  history: mockHistory,
-};
+import { generateHistoryMock, generateRoutesMock } from './mocks';
```
This test file was exporting a mock, which was being imported by other tests in a different module. The import itself would cause these test here to also run and i was getting weird errors. So I moved the mocks to a separate module to avoid importing it from a test file

machadoum reviewed (approved):
security-threat-hunting-explore changes LGTM!

michaelolo24 commented on x-pack/plugins/security_solution/public/common/utils/route/mocks.ts:
```diff
@@ -0,0 +1,44 @@
+
+const pop: Action = 'POP';
+
+const generateDefaultLocationMock = () => ({
```
Thanks for making all of these functional btw.

paul-tavares replied:
NP. Yeah, its still unclear to me what was causing the weird failures and why they started happening now. What made it even harder to debug was that if I ran the single test that failed locally, it would run successful. it was only when I ran the entire suite (security solution) that I was getting failures, so I started to remove the shared/mutable mocks between tests suites and that seemed to have done it.

michaelolo24 commented:
Really awesome feature. Looks good, and doesn't seem like anything is affected when the feature flag is off. One thing to note (this always comes up for timeline), but the z-index will need to be raised for it to show in the timeline context: (See about 25 seconds in the below video)

https://user-images.githubusercontent.com/17211684/170102369-628ef90b-d206-449e-b5a6-37ece8f7701f.mov

michaelolo24 reviewed (approved):
Thanks for this awesome feature! Just a note to update the z-index before the feature flag is default "on". Thanks LGTM!

paul-tavares commented on x-pack/plugins/security_solution/public/detections/components/response_actions_console/response_actions_console_context_menu_item.tsx:
```diff
@@ -0,0 +1,89 @@
+      error,
+    } = useGetEndpointDetails(endpointId, { enabled: Boolean(endpointId) });
+
+    const [isDisabled, tooltip]: [disabled: boolean, tooltip: ReactNode] = useMemo(() => {
```
Good point. I'll make the change in my next PR. 👍

paul-tavares commented:
Thanks @michaelolo24 . Appreciate the time you took to look it through.
